File: backend/src/main/java/com/backend/domain/question/entity/Question.java
package com.backend.domain.question.entity;

import com.backend.domain.answer.entity.Answer;
import com.backend.domain.user.entity.User;
import com.backend.global.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.*;

import java.util.List;
import java.util.UUID;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Builder
@AllArgsConstructor
public class Question extends BaseEntity {

    @Column(nullable = false, length = 100)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(nullable = true)
    @Builder.Default
    private Boolean isApproved = false; // 0: 미승인, 1: 승인

    @Column(nullable = false)
    @Builder.Default
    private Integer score = 0; // 기본값 0

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User author;


    @Enumerated(EnumType.STRING)
    @Column(nullable = true)
    private QuestionCategoryType categoryType;

    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY, cascade = CascadeType.REMOVE)
    private List<Answer> answers;

    @Column(name = "group_id", columnDefinition = "BINARY(16)")
    private UUID groupId;

    public void updateApproved(Boolean isApproved) {
        this.isApproved = isApproved;
    }

    public void updateScore(Integer newScore) { this.score = newScore; }

    public void updateUserQuestion(String title, String content, QuestionCategoryType categoryType) {
        this.title = title;
        this.content = content;
        this.categoryType = categoryType;
    }

    public void updateAdminQuestion(String title, String content, Boolean isApproved, Integer score, QuestionCategoryType categoryType) {
        this.title = title;
        this.content = content;
        this.categoryType = categoryType;
        updateApproved(isApproved);
        updateScore(score);
    }

    public void changeCategory(QuestionCategoryType categoryType) {
        this.categoryType = categoryType;
    }

}

File: backend/src/main/java/com/backend/domain/question/entity/Question.kt
package com.backend.domain.question.entity

import com.backend.domain.answer.entity.Answer
import com.backend.domain.user.entity.User
import com.backend.global.entity.BaseEntity
import jakarta.persistence.*
import java.util.*

@Entity
@Table(name = "question")
open class Question(

    @Column(nullable = false, length = 100)
    var title: String,

    @Column(nullable = false, columnDefinition = "TEXT")
    var content: String,

    @Column(nullable = true)
    var isApproved: Boolean = false,

    @Column(nullable = false)
    var score: Int = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    var author: User,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var categoryType: QuestionCategoryType,

    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY, cascade = [CascadeType.REMOVE])
    var answers: MutableList<Answer> = mutableListOf(),

    @Column(name = "group_id", columnDefinition = "BINARY(16)")
    var groupId: UUID? = null

) : BaseEntity() {

    // JPA 기본 생성자
    protected constructor() : this(
        title = "",
        content = "",
        isApproved = false,
        score = 0,
        author = User(),
        categoryType = QuestionCategoryType.DATABASE,
        answers = mutableListOf(),
        groupId = null
    )

    fun updateApproved(isApproved: Boolean) {
        this.isApproved = isApproved
    }

    fun updateScore(newScore: Int) {
        this.score = newScore
    }

    fun updateUserQuestion(title: String, content: String, categoryType: QuestionCategoryType) {
        this.title = title
        this.content = content
        this.categoryType = categoryType
    }

    fun updateAdminQuestion(
        title: String,
        content: String,
        isApproved: Boolean,
        score: Int?,
        categoryType: QuestionCategoryType
    ) {
        this.title = title
        this.content = content
        updateApproved(isApproved)
        score?.let { updateScore(it) }
        this.categoryType = categoryType
    }

    fun changeCategory(categoryType: QuestionCategoryType) {
        this.categoryType = categoryType
    }

    // Builder
    class Builder {
        private var title: String = ""
        private var content: String = ""
        private var author: User = User()
        private var categoryType: QuestionCategoryType = QuestionCategoryType.DATABASE
        private var isApproved: Boolean = false
        private var score: Int = 0
        private var groupId: UUID? = null

        fun title(title: String) = apply { this.title = title }
        fun content(content: String) = apply { this.content = content }
        fun author(author: User) = apply { this.author = author }
        fun categoryType(categoryType: QuestionCategoryType) = apply { this.categoryType = categoryType }
        fun isApproved(isApproved: Boolean) = apply { this.isApproved = isApproved }
        fun score(score: Int) = apply { this.score = score }
        fun groupId(groupId: UUID) = apply { this.groupId = groupId }

        fun build(): Question {
            return Question(
                title = title,
                content = content,
                isApproved = isApproved,
                score = score,
                author = author,
                categoryType = categoryType,
                groupId = groupId
            )
        }
    }

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }
}